src/main/user-impl/java/com/mysql/cj/jdbc/ha/plugins/ReadWriteSplittingPlugin.java

```diff
@@ -0,0 +1,271 @@
+import java.util.List;
+import java.util.concurrent.Callable;
+
+public class ReadWriteSplittingPlugin implements IConnectionPlugin {
+  public static final int WRITER_INDEX = 0;
+  public static final String METHOD_SET_READ_ONLY = "setReadOnly";
+
+  private final ICurrentConnectionProvider currentConnectionProvider;
+  private final IConnectionProvider connectionProvider;
+  private final PropertySet propertySet;
+  private final IConnectionPlugin nextPlugin;
+  private final Log logger;
+
+  private boolean inTransaction = false;
+  private boolean readOnly = false;
+  private List<HostInfo> hosts = new ArrayList<>();
+  private JdbcConnection writerConnection;
+  private JdbcConnection readerConnection;
+
+  public ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this(currentConnectionProvider, new BasicConnectionProvider(), propertySet, nextPlugin, logger);
+  }
+
+  ReadWriteSplittingPlugin(
+      ICurrentConnectionProvider currentConnectionProvider,
+      IConnectionProvider connectionProvider,
+      PropertySet propertySet,
+      IConnectionPlugin nextPlugin,
+      Log logger) {
+    this.currentConnectionProvider = currentConnectionProvider;
+    this.connectionProvider = connectionProvider;
+    this.propertySet = propertySet;
+    this.nextPlugin = nextPlugin;
+    this.logger = logger;
+  }
+
+  @Override
+  public Object execute(Class<?> methodInvokeOn, String methodName, Callable<?> executeSqlFunc, Object[] args) throws Exception {
+    if(METHOD_SET_READ_ONLY.equals(methodName)) {
+      switchConnectionIfRequired((boolean) args[0]);
+    }
+    return this.nextPlugin.execute(methodInvokeOn, methodName, executeSqlFunc, args);
+  }
+
+  void switchConnectionIfRequired(boolean readOnly) throws SQLException {
+    JdbcConnection currentConnection = this.currentConnectionProvider.getCurrentConnection();
+    if (readOnly) {
+      if (!isReaderConnection() || currentConnection.isClosed()) {
+        try {
+          switchToReaderConnection(currentConnection);
+        } catch (SQLException e) {
+          if(currentConnection == null || currentConnection.isClosed()) {
+            throw new SQLException(
+                Messages.getString("ReadWriteSplittingPlugin.1"),
+                MysqlErrorNumbers.SQL_STATE_UNABLE_TO_CONNECT_TO_DATASOURCE, e);
+          }
+
+          // stick with the current connection
+          if (this.readerConnection != currentConnection) {
+            if (this.readerConnection != null) {
+              this.readerConnection.close();
+            }
+            this.readerConnection = currentConnection;
+          }
+        }
+      }
+    } else {
+      if (!isWriterConnection() || currentConnection.isClosed()) {
+        try {
+          switchToWriterConnection(currentConnection);
+        } catch (SQLException e) {
+          throw new SQLException(
```

Comment:
Do we need to check if connection is open or not, similar to line 66?

Reply:
I don't think so. In the line 66 scenario we have attempted to switch to a reader connection but failed, so I perform the check to see if we can just use the current connection as a fallback. This should be fine since read operations can be sent to reader or writer instances. However, in the scenario on this line, we have attempted to connect to the writer but failed. We don't know if the current connection is to a reader or writer, so we cannot use it and have to throw the exception. Let me know if I'm missing something though